### PR TITLE
Write an error if argument is a directory in Get-FileHash cmdlet

### LINKED ISSUE #11110: Get-FileHash access denied
When using PS 7 preview 5 Get-FileHash it returns access denied to path from pipe

ex. 
```
Get-ChildItem -path "C:\Program Files\" -Recurse | Get-FileHash
Get-FileHash: Access to the path 'C:\Program Files\Common Files' is denied.

From Windows PS
Get-ChildItem -path "C:\Program Files\" -Recurse | Get-FileHash
Algorithm       Hash                                                                   Path
---------       ----
```

### PR DESCRIPTION
**Pull request by iSazonov**
# PR Summary

Fix #11110

If Get-FileHash gets a folder as argument we should write non-terminate error.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

**Comment by iSazonov** [comment]
@SteveL-MSFT If the new test uses $error[0]:
```powershell
        It "Should write non-terminating error if argument is a folder" {
            $result = $pshome, "${pshome}\pwsh.dll" | Get-FileHash
            $result.Count | Should -Be 1
            $error[0].FullyQualifiedErrorId | Should -BeExactly "UnauthorizedAccessError,Microsoft.PowerShell.Commands.GetFileHashCommand"
        }
```
it unexpectedly fail with an ConciseView error
```powershell
MethodInvocationException:
Line |
 127 |                                                     $remainingMessage = $remainingMessage.Substring($substring.Length - $prefix.Length).Trim()

     |                                                     ^ Exception calling "Substring" with "1" argument(s): "StartIndex cannot be less than zero. (Parameter 'startIndex')"
```
I cannot find simple repo steps.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetHash.cs [hunk 1]
```diff
@@ -140,12 +140,22 @@ protected override void ProcessRecord()
                 }
                 catch (FileNotFoundException ex)
                 {
-                    ErrorRecord errorRecord = new ErrorRecord(ex,
+                    ErrorRecord errorRecord = new ErrorRecord(
```
Comment (vexx32) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetHash.cs]: Not blocking, but seems a bit repetitive. 🙂 
```suggestion
                    var errorRecord = new ErrorRecord(
```

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetHash.cs [hunk 2]
```diff
@@ -140,12 +140,22 @@ protected override void ProcessRecord()
...
+                        ex,
                         "FileNotFound",
                         ErrorCategory.ObjectNotFound,
                         path);
                     WriteError(errorRecord);
                 }
+                catch (UnauthorizedAccessException ex)
+                {
+                    ErrorRecord errorRecord = new ErrorRecord(
```
Comment (vexx32) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetHash.cs]: Similar thing here too.
```suggestion
                    var errorRecord = new ErrorRecord(
```

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetHash.cs [hunk 3]
```diff
@@ -140,12 +140,22 @@ protected override void ProcessRecord()
...
                 }
+                catch (UnauthorizedAccessException ex)
+                {
+                    var errorRecord = new ErrorRecord(
+                        ex,
+                        "UnauthorizedAccessError",
+                        ErrorCategory.InvalidData,
+                        path);
+                    WriteError(errorRecord);
+                }
```
Comment (vexx32) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetHash.cs]: Do you think it's worth doing a general catch or additional catches here as well? Looking at the exceptions that can be thrown from [`System.IO.File.OpenRead`](https://docs.microsoft.com/en-us/dotnet/api/system.io.file.openread?view=netcore-3.0#exceptions) it seems likely we can run into the following exceptions somewhat commonly as well:

- `PathTooLongException`
- `IOException`

Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetHash.cs]: IOException combines  FileNotFoundException, DirectoryNotFoundException, and PathTooLongException.

I guess PathTooLongException will be raised above at path resolving time.

So nothing to do. :-)

**Review by SteveL-MSFT — APPROVED** [review]

**Comment by JordanNash** [comment]
It would be appreciated if this error could be changed from "Access Denied" to something more useful like "SKIPPED: Invalid path: folder" or something. 

I'm trying to recurse through my photo archive to try to find duplicates. For each file, I send FullName to get-filehash. I get a LOT of these errors, which implies a permissions issues, but after finding this pull request, I see that it is working as expected. 
Error:
`Get-FileHash: Access to the path 'E:\PHOTO\Lightroom CC Library\Lightroom CC\684f1da3ca494b3e854964d66fd6f341\originals\2011\2011-12-22' is denied.`

Invocation:
`Get-ChildItem -Recurse -path E:\PHOTO\ | Select-Object -ExpandProperty FullName | Get-FileHash | Export-Csv -Append h:\photo_hash_list3.csv`

**Comment by vexx32** [comment]
You can add the `-File` param to your Get-ChildItem call to avoid the issue entirely.

But yeah, that's a fair call. Would be a good idea to open a new issue for that. 🙂

**Pull request merged**